### PR TITLE
LG-2488 Merge the reporter and server and make them both ruby code

### PR DESCRIPTION
**Pull request by jmhooper**
This commit takes the 2 parts of the charts server and merges them into 1 unit that allows the charts server to fetch data on its own.

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -0,0 +1,4 @@
+require 'sinatra/base'
+require 'sinatra/activerecord'
+require 'sinatra/activerecord/rake'
+require './lib/app'
```
Comment (zachmargolis) [inline Rakefile:4]: should this be `require_relative` ?

File: app.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  configure :test do
+    ActiveRecord::Base.logger&.level = :warn
+  end
+
+  configure :production do
+    ActiveRecord::Base.logger&.level = :info
+  end
+
+  get '/' do
+    chart_index = Time.now.to_i / 60 % CHARTS.length
```
Comment (zachmargolis) [inline app.rb:24]: should we make sure this is UTC? `Time.now.utc`? Or set up `Time.zone = ActiveSupport::TimeZone['UTC']` if we want to use activesupport

File: db/migrate/20200218161927_create_reports.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class CreateReports < ActiveRecord::Migration[6.0]
+  def change
+    create_table :reports do |t|
+      t.string :name
+       t.json :data
```
Comment (zachmargolis) [inline db/migrate/20200218161927_create_reports.rb]: nit: fix the alignment/indentation plz

File: lib/reporter/authentications_report_builder.rb [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        datum.timestamp >= start_of_day
+      end
+    end
+
+    def cloudwatch_datapoints
+      @cloudwatch_datapoints ||= cloudwatch_client.get_metric_statistics(
+        namespace: 'Authentication',
+        metric_name: 'user-marked-authenticated',
+        start_time: one_week_ago,
+        end_time: Time.now,
```
Comment (zachmargolis) [inline lib/reporter/authentications_report_builder.rb]: ditto: set a timezone here?

File: lib/reporter/authentications_report_builder.rb [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        current_time += 3600
+      end
+      result
+    end
+
+    def start_of_day
+      Time.new(
+        Time.now.year,
+        Time.now.month,
+        Time.now.day,
```
Comment (zachmargolis) [inline lib/reporter/authentications_report_builder.rb]: What about replacing this all with some ActiveSupport niceness:

```ruby
Time.zone.now.beginning_of_day
```

File: lib/reporter/authentications_report_builder.rb [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+      )
+    end
+
+    def one_week_ago
+      @one_week_ago ||= begin
+        time = Time.now - 7 * 24 * 3600
+        Time.at(time.to_i - (time.to_i % 3600)).getlocal # Round down to 1 hour interval
```
Comment (zachmargolis) [inline lib/reporter/authentications_report_builder.rb]: ActiveSupport:

```ruby
1.week.ago.change(hour: 0, minute: 0)
```

File: spec/support/aws/get_metric_statistics_response.rb [hunk 7]
```diff
@@ -0,0 +1,2 @@
+GetMetricStatisticsDatapoint = Struct.new(:timestamp, :sum)
```
Comment (zachmargolis) [inline spec/support/aws/get_metric_statistics_response.rb]: WDYT of moving this to its own file to make it easier to find? match rails autoloading conventions

File: app.rb (hunk 2 above)
Comment (jmhooper) [reply app.rb:24]: Yeah, originally I was trying to pull in ActiveSupport but looks like I'm gonna have to. Once I do that I'll go around and fix the time zones for all of these.

File: spec/support/aws/get_metric_statistics_response.rb (hunk 7 above)
Comment (jmhooper) [reply spec/support/aws/get_metric_statistics_response.rb]: No strong feelings, happy to move it.

File: app.rb (hunk 2 above)
Comment (zachmargolis) [reply app.rb:24]: I agree that most ActiveSupport magic is bloaty, but if we just did `require 'active_support/time'` I think we could get most of the benefits

**Comment by jmhooper** [comment]
@zachmargolis: I found some time to go through and clean this up some more. Pretty sure I got all of your comments.

File: lib/reporter.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+require 'logger'
+
+require_relative './reporter/alb_errors_report_builder'
+require_relative './reporter/authentications_report_builder'
+require_relative './reporter/doc_auth_dropoff_report_builder'
+
+module Reporter
+  def self.start_reporting
+    Thread.new do
+      Time.zone = ActiveSupport::TimeZone.new('Eastern Time (US & Canada)')
```
Comment (zachmargolis) [inline lib/reporter.rb]: What if we just set a constant somewhere on startup that was like `TIME_ZONE`? The `Thread.new` makes me worry about races

File: lib/reporter/authentications_report_builder.rb [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+      end
+      result
+    end
+
+    def start_of_day
+      @start_of_day ||= Time.zone.today.beginning_of_day
+    end
+
+    def one_week_ago
+      @one_week_ago ||= 1.week.ago.change(min: 0)
```
Comment (zachmargolis) [inline lib/reporter/authentications_report_builder.rb:107]: `seconds: 0` too?

File: spec/reporter/authentications_report_builder_spec.rb [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+      report = subject.reports[1]
+      data = report[:data]
+
+      expect(report[:name]).to eq(:authentications_today_by_hour)
+      expect(data.length).to eq(24)
+      expect(data.first[:label]).to eq('00:00')
+      expect(data.last[:label]).to eq('23:00')
+
+      stat1 = get_daily_report_datapoint_for_timestamp(
+        timestamp: Time.new(1993, 3, 18, 9, 0, 0, 14_400), data: data,
```
Comment (zachmargolis) [inline spec/reporter/authentications_report_builder_spec.rb]: I think that `Time.parse("1993-3-18T09:00:00-05:00")` is slightly easier to read? But it's also a lot of tests to change

File: spec/support/fixtures/doc-auth-drop-offs-report.json [hunk 11]
```diff
@@ -0,0 +1,46 @@
+10-10-2019 <= date user starts doc auth < 10-24-2019
```
Comment (zachmargolis) [inline spec/support/fixtures/doc-auth-drop-offs-report.json]: The extension to this file is misleading? This is definitely not JSON

**Review by zachmargolis — APPROVED** [review]
LGTM

File: spec/support/fixtures/doc-auth-drop-offs-report.json (hunk 11 above)
Comment (jmhooper) [reply spec/support/fixtures/doc-auth-drop-offs-report.json]: Good catch

File: lib/reporter/authentications_report_builder.rb (hunk 9 above)
Comment (jmhooper) [reply lib/reporter/authentications_report_builder.rb:107]: I was messing around with this and it looks like 0ing out minutes also 0ed out seconds

Comment (jmhooper) [reply lib/reporter/authentications_report_builder.rb:107]:
```
irb(main):003:0> 1.week.ago.change(min: 0)
=> 2020-03-06 12:00:00 -0500
```

File: lib/reporter.rb (hunk 8 above)
Comment (jmhooper) [reply lib/reporter.rb]: Good call, I'll do that

File: spec/reporter/authentications_report_builder_spec.rb (hunk 10 above)
Comment (jmhooper) [reply spec/reporter/authentications_report_builder_spec.rb]: Lots of tests but worth it imo, willfix

Comment (zachmargolis) [reply spec/reporter/authentications_report_builder_spec.rb]: bless 🙏

File: lib/reporter/authentications_report_builder.rb (hunk 9 above)
Comment (zachmargolis) [reply lib/reporter/authentications_report_builder.rb:107]: Neat! TIL 💡

**Pull request merged**